### ISSUE TITLE
Flicker When Opening A File/Diff From A PR

### ISSUE THREAD

**Issue by hediet** [microsoft/vscode#185551]
![JdXPsNjOiH](https://github.com/microsoft/vscode-pull-request-github/assets/2931520/fb7579df-b3e9-457b-8c0c-f0edfc57d2e8)

Here the frame where it happens:

![azbB4B0uoK](https://github.com/microsoft/vscode-pull-request-github/assets/2931520/212c23ba-f372-4519-bcc8-d066714eb630)

I'm pretty sure this is caused by the comments gutter.
Is there a way that we can get rid of this flickering?

**Comment by alexr00** [comment]
> I think the extension could set a context key with all the files that have this gutter, so the editor knows immediately whether to reserve some space or not.

This feels like a hack. Here's the existing API:

https://github.com/microsoft/vscode/blob/ba8caef810dff8f38daae1b5465d1bc47a5445fc/src/vscode-dts/vscode.d.ts#L15611-L15616

We could do something similar to the tree data provider to signal that commenting ranges have changed: 

https://github.com/microsoft/vscode/blob/ba8caef810dff8f38daae1b5465d1bc47a5445fc/src/vscode-dts/vscode.d.ts#L10742-L10747

So we'd have something like this:
```ts
	/**
	 * Commenting range provider for a {@link CommentController comment controller}.
	 */
	export interface CommentingRangeProvider {
		/**
		 * Provide a list of ranges which allow new comment threads creation or null for a given document
		 */
		provideCommentingRanges(document: TextDocument, token: CancellationToken): ProviderResult<Range[]>;

		/**
		 * Alert the editor that the list files that contain commenting ranges has changed.
		 */
		onDidChangeCommentingRanges: Event<{resources: Uri[]}>;
	}
```

**Comment by hediet** [comment]
But how would the editor now synchronously before opening a document that there are commenting ranges in it?

**Comment by alexr00** [comment]
The `CommentingRangeProvider` would alert VS Code core by firing `onDidChangeCommentingRanges` as soon as it knows which files have ranges. Then when the editor is opened we will already have a list of files that have ranges.

**Comment by hediet** [comment]
>  Alert the editor that the list files that contain commenting ranges has changed.

But I guess it would not report what changed, but just the new list of resources.

**Comment by alexr00** [comment]
Correct, just the list of resources. This should be cheaper for the extension and will provide enough info for us to reserve the gutter space to prevent the flicker.

**Comment by hediet** [comment]
This would still be nice. Let me know if you need help!

**Comment by alexr00** [comment]
Proposal:

https://github.com/microsoft/vscode/blob/03bd0bb8d117dbe4ea13da19b03b82e8b70194a8/src/vscode-dts/vscode.proposed.commentingRangeResourcesChanged.d.ts#L8-L10

**Comment by alexr00** [comment]
New proposal:
Schemes can be added with this API proposal:

https://github.com/microsoft/vscode/blob/0ae498f89dcf70d8a642cb09af589ccfa19540d9/src/vscode-dts/vscode.proposed.commentingRangeHint.d.ts#L13-L15

For individual resources, extension can now (as of https://github.com/microsoft/vscode/pull/206421) call `vscode.workspace.openTextDocument` early for the files that will have commenting ranges, which will cause VS Code pre-fetch the commenting ranges.

**Comment by alexr00** [comment]
Work is done, moving out of the milestone to finalize at a later date.

**Comment by hediet** [comment]
Works nicely, thank you!
